src/workload.rs

```diff
@@ -115,62 +158,289 @@ impl Workload for SingleInsertion {
+                send!(log, conn, update.as_str())?;
+                send!(log, conn, "commit")?;
+                Ok(())
+            }
+            .await;
+
             if res.is_err() {
-                println!("execute: rollback");
-                conn.query_drop("ROLLBACK")?;
+                send!(log, conn, "rollback")?;
             }
-            println!("{:?}", res);
+            info!(log, "workload finished"; "result" => ?res);
+
+            collect_result(res, results, &table, injection, pool.clone()).await;
+            disable_failpoint(
+                &log,
+                client,
+                "github.com/pingcap/tidb/table/tables/corruptMutations",
+            )
+            .await?;
+        }
 
-            collect_result(res, results, &table, injection, conn);
+        Ok(())
+    }
+}
+
+// similar to T2, but add a deletion after the update.
+struct T3;
+#[async_trait]
+impl Workload for T3 {
+    async fn execute(
+        &self,
+        log: Logger,
+        config: &Config,
+        table: Table,
+        client: &reqwest::Client,
+        pool: Arc<Pool<MySql>>,
+        results: &mut HashMap<(Table, String, String), Effectiveness>,
+    ) -> Result<()> {
+        let mut conn = pool.acquire().await?;
+        enable_featuers(&mut conn, config).await?;
+        let drop_statement = table.drop_statement();
+        let create_statement = table.create_statement();
+        let start = Instant::now();
+        let duration = start.elapsed();
+        CREATE_TABLE_DURAION_MS.fetch_add(duration.as_millis() as u64, Ordering::SeqCst);
+
+        for injection in AVAILABLE_INJECTIONS {
+            send!(log, conn, drop_statement.as_str()).expect("don't let drop statement fail");
+            send!(log, conn, create_statement.as_str()).expect("don't let create statement fail");
+            info!(log, "{} ready to go!", injection);
+
+            // NOTE: "1*" here, otherwise an index mutation is missing for each row insertion, thus cannot be detected.
+            enable_failpoint(
+                &log,
+                client,
+                "github.com/pingcap/tidb/table/tables/corruptMutations",
+                format!("1*return(\"{}\")", injection),
+            )
+            .await?;
+            send!(log, conn, "begin optimistic")?;
+            let row = table.new_row();
+            let insertion = format!("INSERT INTO {} VALUES ({})", table.name, row.to_string());
+            let update = format!(
+                "UPDATE {} SET {} = {} WHERE {} = {}",
+                table.name,
+                table.cols[0].name,
+                row.cols[0].next().to_string(),
+                table.cols[1].name,
+                row.cols[1].to_string()
+            );
+            let deletion = format!(
+                "DELETE FROM {} WHERE {} = {}",
+                table.name,
+                table.cols[0].name,
+                row.cols[0].to_string()
```

Comment:
Is it expected to delete the row by the old value of cols[0], which seems should not exist? Consider describe your purpose in comments

Reply:
It might be intentional when I write it but yeah deleting using new values is more realisitc, I will change it.